crates/lace-bin/src/typechecking/check_pass/check.rs

```diff
@@ -0,0 +1,564 @@
+                                .metadata,
+                        ),
+                        else_return_span: ctx.span_for(
+                            r.value
+                                .last
+                                .as_ref()
+                                .unwrap_or_lice("should have a return expr")
+                                .metadata,
+                        ),
+                        if_ty: lty.to_string(),
+                        else_ty: rty.to_string(),
+                        if_block_span: ctx.span_for(l.metadata),
+                    });
+                }
+            } else {
+                let lvar = ctx.fresh();
+                typecheck_block(
+                    l,
+                    ctx,
+                    &ReturnContext {
+                        expected_type: lvar.into(),
+                        function_block: false,
+                        ..rctx
+                    },
+                )?;
+
+                if let Err((lty, _)) = ctx.unify(lvar, PartialType::Unit) {
+                    return Err(TypeError::IfWithUnexpectedReturn {
+                        return_span: ctx.span_for(
+                            l.value
+                                .last
+                                .as_ref()
+                                .unwrap_or_lice("should have a return expr")
+                                .metadata,
+                        ),
+                        return_ty: lty.to_string(),
+                        if_block_span: ctx.span_for(l.metadata),
+                    });
+                }
+            }
+
+            Ok(())
+        }
+        ExprKind::Block(b) => typecheck_block(
+            b,
+            ctx,
+            &ReturnContext {
+                block_return_type: rctx.expected_type,
+                function_block: false,
+                ..rctx
+            },
+        ),
+        ExprKind::Ident(v) => {
+            let var = ctx.type_variable_for_identifier(v);
+
+            if let Err((l, r)) = ctx.unify(rctx.expected_type, var) {
+                return Err(TypeError::FailedUnification(FailedUnification {
+                    expected: l.to_string(),
+                    was: r.to_string(),
+                    was_span: ctx.span_for(v.metadata),
+                }));
+            }
+
+            Ok(())
+        }
+        ExprKind::Paren(e) => typecheck_expr(e, ctx, rctx),
+        ExprKind::BinaryOp(op, l, r) => {
+            macro_rules! failed {
+                ($side: literal, $ty: expr) => {
+                    |uni: FailedUnification| TypeError::BinaryOp {
+                        op: op.value,
+                        value_ty: uni.was,
+                        expected_ty: $ty.to_string(),
+                        side: $side,
+                        value_span: uni.was_span,
+                        op_span: ctx.span_for(op.metadata),
+                    }
+                };
+            }
```

Comment:
Surely this can be written without a macro. It could maybe even be a closure.

Reply:
nope, lifetimes are difficult, I did have it a closure first but it's unexpectedly hard.